[xavierpok] iP

xavierpok opened the pull request:
# **Kniaz**

> I've invented a new spinny thing! I think I shall call it a "weel" -Me, being unsubtle
For similar products, look [here](https://www.google.com/search?q=task+list+app&oq=task+list+app&aqs=chrome..69i57j0i512l9.6575j0j7&sourceid=chrome&ie=UTF-8&bshm=rime/1)

## New Features

- More commands
- Cleaned up design -- Each command has its own `CommandHandler`
- Extensible
- Command-Line Interface

## How to call from another java program 

```java
    public static void main(String[] args) {
        KniazSession.init().run();
    }
```

## Feature map

- [x] Making tasks
- [x] Marking/unmarking tasks as done
- [x] Associating certain tasks with date/times
- [x] Deleting tasks
- [x] Searching for tasks
- [ ] GUI

### Naming explanation

1. Kniaz is a slavic equivalent of Duke, apparently
2. I thought it sounded cool 😮

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

xavierpok committed:
Implements level 1 features & adds (public) constant fields.

Level 1 features just refers to echoing any input given by the user. Also adds a public field EXITSTRING which is the input that will cause Kniaz to exit. Adds private field EXITMESSAGE that is the message that prints on exit.

xavierpok committed:
Implements Level 0 features of the Duke (Re-named to Kniaz for this version) chatbot. This refers to just renaming the chatbot, and implementing a skeletal version that greets the user and exits.

Adds two new private fields, LOGO and SEPERATOR, to help with this, which hold the logo and horizontal seperator lines respectively (Seperator lines are to help with formatting)

xavierpok committed:
Adds  ability to store text entered by user and display back when requested

When a string that isn't interpreted as a command is input, adds it to internal storage. When list is input, displays the list. Note : 'bye' is retained as a command still.

Also adds a new private function to format this task list as a string.

xavierpok committed:
Final version of Level 2

Makes edits to code formatting, functionality is the same. Redundant while statement amended to while(True), and System.out.println of String.format call amended to System.out.printf

xavierpok committed:
Adds ability to mark/unmark tasks as done/undone with accompanying message.
Adds Task and TaskList classes to encapsulate tasks, and lists of tasks respectively.

xavierpok committed:
Documents additions in previous commit, tagged Proto-Level-3

xavierpok committed:
Implements ToDos, Deadlines and Events, as well as input/output (i.e. commands and printing to Kniaz's command line) for said tasks, and refactors Task printing slightly.

Feature expansion to hit Level 4.

ToDos are Tasks without dates/times, Deadlines are Tasks with a singular due date/times, Events are Tasks with a start/stop.

Also refactors Task to be an abstract class for ToDo to inherit from.

This is a slight shift of paradigm, as a Task would be something you need to do, but alone contains insufficient information.

ToDos lack time as it's just something the user has to do -- time isn't relevant here

A deadline naturally has a due date.

Events in Kniaz are defined by start-stop in addition to the name, which is a natural abstraction of an event (Kniaz doesn't care about place (yet?))

Lastly, string representation to output of Tasks in Kniaz's output has been shifted from toString() to toPrintString(), to diffrentiate between input meant SOLELY for the user (which is toPrintString())

xavierpok committed:
Patches user-facing string rep of Event, adding a whitespace in front of "from"

Needed to make it adhere closer to expected output, as no whitespace means "from" sticks to the previous string

Improves output seen by user

xavierpok committed:
Patches user-facing string rep of Event, adding a whitespace in front of "from"

Needed to make it adhere closer to expected output, as no whitespace means "from" sticks to the previous string

Improves output seen by user

xavierpok committed:
Merge remote-tracking branch 'origin/master'

xavierpok committed:
Adds (single) semi-automated test case for use with runtest.sh

Allows for quick verification of input/output, comparing EXPECTED against input and actual output.

Expected is to be updated with further edits to behaviour of Kniaz.

xavierpok committed:
Adds user error reporting, moves a significant amount of parsing and event logic out of Kniaz, and adds a new Exception, KniazRuntimeException

User error reporting is useful, as the user would naturally want to know what they did wrong.

Moves parsing of commands into KniazParser, which now handles parsing of commands. It seperates commands into what kind of commands and their arguments

Moves logic of commands into the logic.taskhandling package. The classes within take in the text-interface command arguments as Strings and work through the logic of how to operate on those. E.g., the ToDo name is passed to a ToDoHandler which decides what to do with this.

Moving of parsing helps to keep Kniaz clean, and seperate how we decide to parse the command given to Kniaz seperate from the main Kniaz loop. This is important as there are many ways to parse arguments, and we might decide another way is better. This also allows for better exception handling, as Kniaz does not need to do input validation, and just passes it off to another class. All Kniaz has to do now is to grab whatever exception is thrown up and figure out what to do(Which is a lot easier!)

Similarly, for logic, we might decide to refactor it later, and this helps keep Kniaz clean. Notably, validation of the arguments provided can be handled seperately. For example, ToDo will only expect ONE argument, but the user might decide to provide two. Getting Kniaz or KniazParser to handle this borders on scope creep of what these classes are meant to do. Also offers similar benefits in exception handling as per the paragraph above about KniazParser.

Addition of the new Exception is to support these new features. Notably, the addition of a seperate userMessage field supports two different exception reporting pipelines, to support good user experience. It's reasonable that more information might be needed to debug for future feature additions/refactors that might overwhelm or confuse the user, but a developer might find useful.

Modifies test cases to reflect the new change in behaviour

xavierpok committed:
Documents code added in previous commit(Tagged Level-5), Converts all commandHandler classes in logic->taskhandling package to be abstract classes

Documentation was not done previously, this commit fixes that

Classes should be abstract, as they contain only static methods and should not be instantiated.

xavierpok committed:
Updates test case

Amends and expands input.txt and EXPECTED.txt to be accurate to new features, including invalid input handling

xavierpok committed:
Implements task deletion, and fixes formatting & documentation in TaskList, and adds whitespace stripping to parseCommand

DeleteHandler follows the same design pattern for the other commands, and additions in Kniaz and KniazParser are needed to accomodate the new command.

Adding deletion is a useful feature because sometimes the user makes a mistake or wants to permanently delete a task to shorten the list

Also adds test case for demonstration of new behaviour

parseCommand stripping whitespace brings it closer to user-expected behaviour because it's natural to expect that leading/trailing whitespaces would be ignored

Lastly, fixing of formatting and adding documentation is good because style guide and it's good to document

xavierpok committed:
Implement saving

Currently, Kniaz forgets all data on new session

Let it save to remember from previous session.

xavierpok committed:
Implement loading on Kniaz start

Previous commit tagged Saving only saves, does not load

Saving only makes sense if Kniaz loads saved data

xavierpok committed:
Amend test case

New changes in commits tagged Loading and Saving modify output behaviour

Update expected to match

xavierpok committed:
Write javadoc strings for commits tagged Saving, Loading

Previous commits implemented functionality, lacked javadoc strings.

Documentation is good.

xavierpok committed:
Merge branch 'branch-level-7'

xavierpok committed:
Implement date-time parsing

Currently, stores date-time only as Strings, no parsing

Should parse to allow for flexibility and organised output for better UX

xavierpok committed:
Write javadoc strings & fix code standard issues

Documentation & code standards are important

Modify access on class members to keep the API good and upkeep good OOP

xavierpok committed:
Update test cases

Test cases were unmodified, outdated

Modification to reflect new behaviour is good

xavierpok committed:
Merge branch 'branch-level-8'

xavierpok committed:
Create new package io

Prepare to seperate I/O into another package for better OOP practices

xavierpok committed:
Move parser package to io & rename

Was not under io package previously

Parsing of incoming inputs should be under io package

xavierpok committed:
Seperate InstructionType into own class

Was inner class of KniazParser

Now, seperate to make cleaner, needed as more functionality added on

xavierpok committed:
Create exception class shells

One type of exception covers all Kniaz exception, very unspecific

Expand types to prepare for I/O refactor later

xavierpok committed:
Refactor command handling & parsing

Command handling code used to handle parsing also, and was messy.

Split command handling & parsing to be more functional in paradigm.

Better design helps maintain and extend

xavierpok committed:
Finish implementation and add javadocs

Had some small errors and unfinished docs

Should be cleaned up and documentation thrown in for future reference

xavierpok committed:
Fix error in CommandFactory, KniazCommand

Now, does not make sense or works

Sensible and working code is good

xavierpok committed:
Fix errors, finish implementation of OOP

Previously was undone, finish refactor

Better methodology and structure is good for design

Current implementation lacking exception handling, output is bare-bones

xavierpok committed:
Make chatbot have more personality in responses

Previous refactor removed due to re-implementation of I/O

Feature re-addition after refactor is good to return to behaviour

Current design also more extensible overall than previous design

xavierpok committed:
Add missing javadoc strings & inline comments

Previously, missing javadoc strings & inline comments in code.

Consistent documentation is a good thing.

xavierpok committed:
Fix bugs in line parsing

Previously, read each whitespace delimited string as an unnamed argument

I.e. it would interpret "todo read book" w/ args "read"&"book"

Fix this to be in line with old & expected behaviour

xavierpok committed:
Update exceptions, use & handle new exceptions

Fine-tune exception types and implement where appropriate

Better practices, as more specific exceptions are better

When exception is thrown, more information about what kind of exception

Previous commits did not handle exception after refactor

Handling certain exceptions that are expected to crop up is good

xavierpok committed:
Update test case to match expected behaviour

Changes changed behaviour slightly, update to match expected output

Updating test case is good because now it reflects actual expected

xavierpok committed:
Add comments explaining why things were done

Documentation is good

xavierpok committed:
Removes unused import statements and fixes javadoc

Code style and documentation standards are good

xavierpok committed:
Merge branch 'branch-A-MoreOOP'

xavierpok committed:
Merge branch 'add-gradle-support'

xavierpok committed:
Add & configure gradle

xavierpok committed:
Add no-arg constructor to KniazSession

Useful for instantiating anonymous instances -- Lack all behaviour as no fields

xavierpok committed:
Add tests for DeadlineHandler, DeleteHandler

xavierpok committed:
Merge branch 'branch-A-JUnit'

xavierpok committed:
Add new find command

Feature expansion, useful if user has many tasks in their task list.

xavierpok committed:
Edit name of built jar via shadowjar to Kniaz

Kniaz is the name of our program, NOT duke.

xavierpok committed:
Merge branch 'Branch-L-9'

prawnzyy commented on src/main/java/ui/inputparser/KniazLineParser.java:
```diff
@@ -0,0 +1,122 @@
+import main.logic.command.CommandFactory;
+import main.logic.command.KniazCommand;
+
+import java.util.*;
```
I think try to import the exact class and avoid doing star import so as to make it clearer what class you are using?

prawnzyy commented on src/main/java/ui/inputparser/InstructionType.java:
```diff
@@ -0,0 +1,64 @@
+    LIST("list", 0, 0, new String[]{}),
+    QUIT("bye", 0 , 0, new String[]{}),
+    DELETE("delete", 1, 0, new String[]{}),
+    FIND("find",1,0,new String[]{}),
```
Try to leave a space after a comma just like your above statements

xavierpok replied:
Was a new addition, didn't notice the variance in style, will fix!

prawnzyy commented on src/main/java/main/logic/command/CommandFactory.java:
```diff
@@ -0,0 +1,54 @@
+package main.logic.command;
+
+
+import main.logic.handler.*;
```
Try to import the direct class/classes rather than using star import

xavierpok replied:
Right, missed this, thanks!

prawnzyy reviewed (commented):
Overall, I found your code to be well split and the classes are well defined. There are quite a few chunks of code commented out. If they are no longer in use, try to remove them to improve readability. Also take note of your spacing and line breaks as they are inconsistent across each file. I pointed out a few small nits but other than that, LGTM!

xavierpok committed:
Fix codestyle errors & import statements

xavierpok committed:
Fix to adhere to stylecheck

xavierpok committed:
Fix format bug that caused compile error

xavierpok committed:
Fix compile error

xavierpok committed:
Implement first stage of javafx

xavierpok committed:
Revert "Fix compile error"

This reverts commit 5566d57024d61d1f6b9cc54f0280697e67f2aa89.

xavierpok committed:
Merge branch 'master' into javafx-testing

xavierpok committed:
Implement framework from javafx tutorial

xavierpok committed:
Re-work Kniaz to use javaFX GUI to I/O

Part of implementation of Level-10

xavierpok committed:
Implement quitting

xavierpok committed:
Add javadocs

xavierpok committed:
Refactor, mostly sorting into packages

xavierpok committed:
Merge branch 'branch-Level-10'

jacksonyuan256 commented on src/main/java/storage/TaskList.java:
```diff
@@ -0,0 +1,68 @@
+ * Class encapsulating a list of Tasks, supports easy string representation, refer to Task for more details about
+ * what Tasks are.
+ */
+public class TaskList extends ArrayList<Task> implements Serializable {
```
Interesting use of inheritance here.

jacksonyuan256 commented on src/main/java/main/logic/command/CommandFactory.java:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Class encapsulating a command given to Kniaz, should only be instantiated via KniazParser when it parses a command
+ */
+public abstract class CommandFactory {
+    private static final EnumMap<InstructionType, CommandHandler> INSTRUCT_TO_HANDLER =
+            new EnumMap<>(Map.of(
+                    InstructionType.TODO, new ToDoHandler(),
+                    InstructionType.DEADLINE, new DeadlineHandler(),
+                    InstructionType.EVENT, new EventHandler(),
+                    InstructionType.MARK, new MarkHandler(),
+                    InstructionType.UNMARK, new UnmarkHandler(),
+                    InstructionType.LIST, new ListHandler(),
+                    InstructionType.QUIT, new QuitHandler(),
+                    InstructionType.DELETE, new DeleteHandler(),
+                    InstructionType.FIND, new FindHandler(),
+                    InstructionType.INVALID, new InvalidHandler()
+            ));
```
I can see a lot of effort put into building a more complex command structure, as well as the mapping of instructions to the abstracted classes. Good job!

jacksonyuan256 commented on src/main/java/storage/save/KniazLoader.java:
```diff
@@ -0,0 +1,95 @@
+            throw new KniazRuntimeException("Savefile does not exist",
+                    "There is no saved data of your tasklist. We shall start from scratch.",
+                    new IOException(String.format("%s does not exist", this.saveFile.getAbsolutePath())));
+            // should not normally be thrown because we verify up above, but to satisfy compiler requirements
```
Yes I was slightly annoyed by this too

jacksonyuan256 commented on src/main/java/storage/save/KniazLoader.java:
```diff
@@ -0,0 +1,95 @@
+     * @return the loaded TaskList
+     * @throws  KniazRuntimeException when something goes wrong in retrieving data, will wrap around the root exception.
+     */
+    public TaskList load() throws KniazRuntimeException {
```
I find it interesting that your architecture involves TaskList being an ArrayList itself and the use of streams itself, which further increases the complexity.

jacksonyuan256 commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,71 @@
+     * whether this task is done and its name
+     * @return the user-facing string representation of this ToDo.
+     */
+    public String toPrintString() {
+        String statusIcon = "";
+        if (this.isDone) {
+            statusIcon = "X";
+        } else if (!this.isDone) {
+            statusIcon = " ";
+        }
+        return String.format("[%s] %s", statusIcon, this.taskName);
+        // return in format [statusIcon] taskname
+    }
```
1 small nitpick is that the output can be dictated by modifying the object toString method, and then you can just return the object itself. But otherwise this is also alright!

jacksonyuan256 reviewed (commented):
Overall very well-done, complex architecture and strictly adhering to OOP. A lot of effort was put in, good job!

xavierpok committed:
Add assert for InstructionType

Currently does not use assert to throw exception.

Changes error checking to use assert.

Using assert makes it cleaner and emphasises it's a programmer error

Exception covered should not occur except in case of error.

xavierpok committed:
Add assert for FindHandler

Currently does not use assert to throw exception.

Changes error checking to use assert.

Using assert makes it cleaner and emphasises it's a programmer error

Exception covered should not occur except in case of error.

It's unlikely to occur but in case something goes wrong.

xavierpok committed:
Add assert for CommandFactory

Currently, does not do checking for instruction types existing in its mapping.

Adds such checking.

Ensures that all instruction types entered by user will have an associated handler.

Exception covered should not occur except in case of error.

xavierpok committed:
Touch up code quality

Bad code is bad.

This makes code better to read.

Commenting documenting strange values & un-nesting code to be better.

xavierpok committed:
Add assert for KniazOutputFlavourer

Currently missing such compile-time checks

Important that all instructiontypes have a flavour string

User will expect Kniaz to give some kind of output for each valid command, this ensures that there is.

Should not happen except in case of programmer error

xavierpok committed:
Merge pull request #1 from xavierpok/A-Assertions

Add assertions to verify compiletime & runtime condiitons at various points in code

xavierpok committed:
Merge pull request #2 from xavierpok/A-CodeQuality

Touch up code quality

xavierpok committed:
Add support for tagging

Currently no such tagging feature.

Adds implementation tagging (without yet letting the user tag)

Tagging allows for later sorting by tag.

xavierpok committed:
Add tagging I/O and touch up tagging code

Currently tagging is not user-accessible.

Adds tagging input/output.

Tagging allows for later sorting by tag.

xavierpok committed:
Add searching for tags in find command

No way to actually use tags

The find command now also finds for tags

Piggyback onto the find command to avoid unneeded command bloat

xavierpok committed:
Merge pull request #3 from xavierpok/C-Tagging

Adds tagging and searching for tagging

xavierpok committed:
Revert "Adds tagging and searching for tagging"

xavierpok committed:
Merge pull request #4 from xavierpok/revert-3-C-Tagging

Revert "Adds tagging and searching for tagging"

xavierpok committed:
Revert "Revert "Adds tagging and searching for tagging""

This reverts commit c358832dc915cb428d0aeb3d417670aa7e32a33c.

There were a few errors, but now they're fixed

xavierpok committed:
Update README.md

Update documentation & add user guide

xavierpok committed:
Add whitespace as dummy commit

Javadocs were done way before, but commit for admin purposes.

xavierpok committed:
Add whitespace as dummy commit

Coding standards were done way before, but commit for admin purposes.

xavierpok committed:
Merge branch 'branch-A-JavaDoc'

xavierpok committed:
Merge branch 'branch-A-CodingStandard'

xavierpok committed:
Change build.gradle settings

Previous build.gradle had Kniaz as the main class

Current main class is Launcher since GUI change

Make functionality functional for .jar build

xavierpok committed:
Add Ui.png

Non-functional change, purely for admin purposes

xavierpok committed:
Edit README.md in docs
